### PR TITLE
Add Stable Diffusion support and cross-platform serialization

### PR DESCRIPTION
**Pull request by harubaru**
I've also implemented a fallback to the `RequestsStreamFile` in case the user is running the tensorizer on a systems that aren't Unix-based since `CURLStreamFile` works with Unix dependencies.

### CODE REVIEW

File: tensorizer.py [hunk 1]
```diff
@@ -3,39 +3,48 @@
...
+import requests
+
+os.environ["TRANSFORMERS_VERBOSITY"] = "error" # disable missing keys and unexpected key warnings
 
 thisPath = str(pathlib.Path(__file__).parent.resolve())
 sys.path.append(thisPath + "/../../transformers/src")
 sys.path.append(thisPath + "/../../")
 
-from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig
+from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig, CLIPTextModel, CLIPTokenizer, CLIPTextConfig
```
Comment (wbrown) [inline tensorizer.py:34]: enclose in `()` and wrap to multiple lines.

File: tensorizer.py [hunk 2]
```diff
@@ -3,39 +3,48 @@
...
 from transformers.modeling_utils import no_init_weights, PreTrainedModel
-from lm_node.unitrim import Unitrimmer
 from tokenizers import Tokenizer
 
+from diffusers import AutoencoderKL, UNet2DConditionModel, StableDiffusionPipeline, LMSDiscreteScheduler
```
Comment (wbrown) [inline tensorizer.py:38]: enclose in `()` and wrap to multiple lines.

File: tensorizer.py [hunk 3]
```diff
@@ -603,10 +535,16 @@ def read_tensors(self) -> Iterator[Tuple[int, int, str, numpy.ndarray]]:
...
-                datalength = struct.unpack("<q", headers[header_len - 8 :])[0]
+                # read the shape amount, according to the serialized format. the shape length is
+                # 1 byte after the dtype end.
+                shape_len = struct.unpack("<B", headers[dtype_end:dtype_end + 1])[0]
+                # the shape elements are <I so we read 4 bytes. _read_shapes takes in the
+                # header object and the number of elements in the shape.
+                # the amount of bytes for the shape is 4 * the number of elements in the shape.
+                # so, we need to read 4 * shape_len bytes after the dtype end + 1 byte for the
+                # shape length. sort of convoluted, but it works.
+                shapelist = self._read_shapes(headers[dtype_end + 1:(dtype_end + 1) + (4 * shape_len)], shape_len)
```
Comment (wbrown) [inline tensorizer.py:546]: Wrap this.

File: tensorizer.py [hunk 4]
```diff
@@ -823,68 +735,173 @@ def serialize_model(m: torch.nn.Module, out_prefix: str):
     # torch.save(y, f"{out_prefix}.state", _use_new_zipfile_serialization=False)
```
Comment (wbrown) [inline tensorizer.py:735]: Remove these commented out lines of code.

File: tensorizer.py [hunk 5]
```diff
@@ -823,68 +735,173 @@ def serialize_model(m: torch.nn.Module, out_prefix: str):
...
+            # We have fcntl, so we can use the fast CURL-based loader.
+            logger.info("Using CURL for tensor streaming")
+            tensor_loader = lambda: CURLStreamFile(tensors_uri)
+        else:
+            # Fallback to slow requests-based loader.
+            logger.info("Using requests for tensor streaming")
+            tensor_loader = lambda: RequestsStreamFile(tensors_uri)
     else:
-        tensors_uri = f"{path_uri}/tensors"
+        tensors_uri = f"{path_uri}/{model_prefix}tensors"
```
Comment (wbrown) [inline tensorizer.py:771]: `.tensors` rather than `tensors`.

File: tensorizer.py [hunk 6]
```diff
@@ -823,68 +735,173 @@ def serialize_model(m: torch.nn.Module, out_prefix: str):
...
         config_uri = "file://" + os.path.join(
-            os.path.dirname(path_uri), "config.json"
+            os.path.dirname(path_uri), f"{model_prefix}config.json"
```
Comment (wbrown) [inline tensorizer.py:773]: `-config.json` rather than `config.json`

File: tensorizer.py [hunk 7]
```diff
@@ -823,68 +735,173 @@ def serialize_model(m: torch.nn.Module, out_prefix: str):
...
+    vae = load_model(output_prefix, AutoencoderKL, None, "vae")
+    unet = load_model(output_prefix, UNet2DConditionModel, None, "unet")
+    encoder = load_model(output_prefix, CLIPTextModel, CLIPTextConfig, "encoder")
+
+    pipeline = StableDiffusionPipeline(
+        text_encoder=encoder,
+        vae=vae,
+        unet=unet,
+        tokenizer=CLIPTokenizer.from_pretrained(sys.argv[1], subfolder="tokenizer"),
+        scheduler = LMSDiscreteScheduler(beta_end=0.012, beta_schedule="scaled_linear", beta_start=0.00085, num_train_timesteps=1000, trained_betas=None),
```
Comment (wbrown) [inline tensorizer.py:857]: Wrap this line.

**Review by wbrown — CHANGES_REQUESTED** [review]
Run this through `black -t 80` to resolve many of the line wrapping issues.

Some other minor comments.

**Review by arsenetar — CHANGES_REQUESTED** [review]
Could we add a docstring or something in the README.md as to what the `hf_main()`, `df_main()`, `goose_main()`  are reference examples for (i.e. there are three different ones why/what is each one showing).  Think that makes them a better reference.

**Comment by harubaru** [comment]
Sure, I can add that in. Should I leave `goose_main()` in also? I don't think it will work with the modifications made.

**Review by wbrown — APPROVED** [review]

**Review by arsenetar — APPROVED** [review]

**Pull request merged**